### PR TITLE
Flakes: wait for the schema tracker to catch up in TestSchemaVersioning

### LINKED ISSUE #21055: Flaky test: TestSchemaVersioning sees the tracker's replayed schema_version rows in its stream
### Overview of the Issue

`go/vt/vttablet/endtoend.TestSchemaVersioning` fails intermittently in the `e2e_race` workflow on `main`. Two recent examples, both on PRs that do not touch vstreamer or the schema tracker:

- https://github.com/vitessio/vitess/actions/runs/34302339048/job/102311724482 (2026-09-09)
- https://github.com/vitessio/vitess/actions/runs/34167695128 (2026-09-07)

The failure is always the first case of the test (`create table vitess_version`), which receives a `VERSION` event where it expected the DDL's `gtid`:

```
--- FAIL: TestSchemaVersioning (0.43s)
    vstreamer_test.go:201:
        Error Trace: .../go/vt/vttablet/endtoend/vstreamer_test.go:479
                     .../go/vt/vttablet/endtoend/vstreamer_test.go:411
                     .../go/vt/vttablet/endtoend/vstreamer_test.go:201
        Error:       Not equal:
                     expected: 1
                     actual  : 17
        Messages:    create table vitess_version (
                     	id1 int,
                     	id2 int
                     ) (2): event: type:VERSION timestamp:1788920379 keyspace:"vttest" shard:"0" commit_parent:660 sequence_number:664 event_gtid:"baf00a98-...:664", want gtid
```

(The `stream (at source tablet) error ... cannot determine table columns for vitess_version` log line that appears in every run of this test, passing or failing, is expected: the test asserts that exact error after an untracked `ALTER`. It is not the failure.)

### Root cause

The test restarts the schema tracker so that it does not "pick events from the previous test":

```go
tsv.SetTracking(false)
...
tsv.SetTracking(true)
time.Sleep(100 * time.Millisecond) // wait for _vt tables to be created
```

and then starts its own VStream from `current` and runs `create table vitess_version`, expecting the only `VERSION` event it sees to be the tracker's `schema_version` row for that DDL.

Since #20279, `Tracker.Open()` resumes from the last position saved in `_vt.schema_version` rather than from `current` (`tracker.go` `startupPosition`). The preceding tests (`views_test.go`) run DDL faster than the tracker saves rows for them, so when `TestSchemaVersioning` calls `SetTracking(false)` the tracker is a few DDLs behind (in one CI run `Close()` also killed the in-flight `insert into _vt.schema_version`, errno 1317). `SetTracking(true)` then resumes from the older saved position and replays those DDLs, writing a `schema_version` row for each. On a slow runner one of those inserts commits *after* the 100 ms sleep, inside the test's stream, ahead of the test's own `create table` — so the stream's third event is that `VERSION` instead of the DDL's `gtid`.

Timeline from run 34302339048 (all inside the test's first ~430 ms):

| time | event |
|---|---|
| 38.765 | `SetTracking(false)`; in-flight `schema_version` insert for DDL @654 killed; last saved position stays `1-650` |
| 39.014 | `SetTracking(true)` → "Schema tracker resuming from the last saved schema version position `1-650`" |
| 39.025–39.036 | tracker replays `vitess_view` DDL @654, @658, @660 from the previous tests |
| 39.115 | 100 ms sleep ends |
| 39.183 | test's VStream starts from `current` = `1-663`; `create table vitess_version` → GTID 665 |
| 39.19x | replayed `schema_version` insert for @660 commits at GTID **664** → test sees `VERSION` before the DDL |

The replay itself is the intended behavior of #20279 (the historian must not miss schema versions across a tracker restart); the bug is that the test's fixed sleep assumes the reopened tracker is idle.

### Reproduction Steps

Removing the sleep makes it fail deterministically (3/3 locally), with the same message as CI:

```sh
# in go/vt/vttablet/endtoend/vstreamer_test.go, delete the
#   time.Sleep(100 * time.Millisecond)
# after tsv.SetTracking(true), then:
go test ./go/vt/vttablet/endtoend/ -race -count=1 \
  -run 'TestCreateViewDDL|TestAlterViewDDL|TestDropViewDDL|TestViewDDLWithInfrSchema|TestViewAndTableUnique|TestSchemaVersioning'
```

The preceding view tests are what leave the tracker behind; `-run TestSchemaVersioning` alone does not reproduce it.

### Binary Version

`main` at d782fbaee6.

### Operating System and Environment details

GitHub Actions `ubuntu-24.04` runners (`e2e_race` workflow); reproduces locally on macOS with the sleep removed.

### Log Fragments

```
schema/tracker.go:105 Schema Tracker: closed
schema/historian.go:98 Historian: opening
schema/tracker.go:79 Schema Tracker: opening
schema/tracker.go:266 Schema tracker resuming from the last saved schema version position position=MySQL56/baf00a98-...:1-650
vstreamer/vstreamer.go:201 Starting Stream() with startPos MySQL56/baf00a98-...:1-650
schema/tracker.go:291 Processing schemaUpdated event for gtid MySQL56/baf00a98-...:1-654, ddl create table vitess_view (
schema/tracker.go:291 Processing schemaUpdated event for gtid MySQL56/baf00a98-...:1-658, ddl drop view if exists vitess_view
schema/tracker.go:291 Processing schemaUpdated event for gtid MySQL56/baf00a98-...:1-660, ddl DROP TABLE IF EXISTS `vitess_view` /* generated by server */
vstreamer/vstreamer.go:201 Starting Stream() with startPos MySQL56/baf00a98-...:1-663
schema/tracker.go:291 Processing schemaUpdated event for gtid MySQL56/baf00a98-...:1-665, ddl create table vitess_version (
endtoend/vstreamer_test.go:178 Received event type:BEGIN ... commit_parent:660 sequence_number:664
endtoend/vstreamer_test.go:178 Received event type:VERSION ... commit_parent:660 sequence_number:664
endtoend/vstreamer_test.go:178 Received event type:GTID gtid:"MySQL56/baf00a98-...:1-664"
endtoend/vstreamer_test.go:178 Received event type:COMMIT ...
endtoend/vstreamer_test.go:178 Received event type:GTID gtid:"MySQL56/baf00a98-...:1-665"
endtoend/vstreamer_test.go:178 Received event type:DDL statement:"create table vitess_version (\n\tid1 int,\n\tid2 int\n)"
--- FAIL: TestSchemaVersioning (0.43s)
```

### PR DESCRIPTION
**Pull request by mattlord**
## Description

`TestSchemaVersioning` has been failing intermittently in `e2e_race` on `main` (twice in the last three days, on PRs that don't touch this code) with:

```
create table vitess_version (...) (2): event: type:VERSION ..., want gtid
```

The test restarts the schema tracker, sleeps 100 ms, starts a VStream from `current`, and expects the only `VERSION` event it sees to be the one for its own `create table`. Since #20279 the reopened tracker resumes from the last `schema_version` row it saved rather than from `current`. The preceding view tests run DDL faster than the tracker saves rows, so it is a few DDLs behind when the test restarts it, and on open it replays them. On a slow runner one of those replayed `schema_version` inserts commits inside the test's stream, ahead of the test's own DDL, and the test sees a `VERSION` where it expected the DDL's `gtid`. The replay is the intended behavior of #20279; the test's fixed sleep was the timing assumption that broke.

The fix replaces the sleep with a barrier: run a DDL and wait for its `schema_version` row. The tracker saves rows in binlog order, so once that row exists the replay is over and the stream only sees the `VERSION` events for this test's DDL. With the sleep removed the test failed 3/3 on `main`; with the barrier (and no sleep) it passes 6/6, and the full `vttablet/endtoend` package passes plain and under `-race`.

Two small supporting changes: the barrier table needs an ACL entry, so the `version` group now matches the `vitess_version` prefix; and `waitForVersionInsert`'s timeout goes from 3 s to 30 s, since it now covers a backlog rather than a single row.

Not a backport candidate: the resume-from-saved-position behavior is only on `main`.

## Related Issue(s)

- Fixes #21055
- Introduced by the test's interaction with #20279

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

## Deployment Notes

None; test-only change.

### AI Disclosure

Investigation and fix written with Claude (Fable) under my direction.

### CODE REVIEW

**Review by migue — APPROVED** [review]

File: go/vt/vttablet/endtoend/vstreamer_test.go [hunk 1]
```diff
@@ -71,7 +71,13 @@ func TestSchemaVersioning(t *testing.T) {
...
 	tsv.EnableHistorian(true)
 	tsv.SetTracking(true)
-	time.Sleep(100 * time.Millisecond) // wait for _vt tables to be created
+	// The reopened tracker resumes from the last schema_version row it saved,
+	// so it may still be replaying DDL from the preceding tests, and each
+	// replayed DDL produces a VERSION event. Wait for it to catch up before
+	// starting the stream below, so that the stream sees only the VERSION
+	// events for this test's own DDL. A fixed sleep raced the replay on slow
+	// runners.
+	waitForSchemaTrackerToCatchUp(t, "create table vitess_version_barrier (\n\tid int\n)")
```
Comment (arthurschreiber) [inline go/vt/vttablet/endtoend/vstreamer_test.go]: I think this is blocking as written: the reopened tracker finding `_vt.schema_version` empty (and taking a fresh snapshot instead of resuming) isn't an edge case, it's the *normal* state for an isolated run.

`servePrimary` → `tracker.Open()` only spawns the goroutine, and `framework.StartCustomServer` returns as soon as `/debug/vars` answers. When this test is the first (or only) test to run, its `SetTracking(false)` cancels the boot-time snapshot mid-flight (`error getting the schema tracker's startup position: context canceled`), so `_vt.schema_version` is empty when `SetTracking(true)` reopens the tracker. The reopened tracker then runs `insertCurrentSchemaSnapshot`, which does a full `engine.Reload` *before* reading the primary position, and that reload is slower than this DDL's round trip. So the snapshot position usually already includes the barrier, the barrier is never streamed, no row ever matches the DDL text, and we sit in `waitForVersionInsert` for 30s.

Repro on this branch, `-race`, `-run '^TestSchemaVersioning$'`, MySQL 8.0.43, three runs:

```
RUN 1  --- PASS (0.71s)
  error getting the schema tracker's startup position: context canceled   <- boot snapshot killed
  Saving schema snapshot for gtid ...:1-73
  Processing schemaUpdated event for gtid ...:1-75, ddl create table vitess_version_barrier   <- landed after the snapshot, got replayed
RUN 2  --- FAIL (30.04s)
  error getting the schema tracker's startup position: Code: UNKNOWN
  Saving schema snapshot for gtid ...:1-73
  Starting Stream() with startPos ...:1-73                                <- barrier already inside 1-73
  schema tracker never saved a schema_version row for create table vitess_version_barrier
RUN 3  --- FAIL (30.06s)   same as run 2
```

CI won't see this because the full package gives the boot snapshot time to land before this test, but anyone reproducing the original flake with `-run` gets a 30s failure more often than not.

Suggestion: don't try to make the barrier itself smarter — I tried a position-based wait first and it doesn't work, because `PrimaryPosition()` read after the DDL also includes the tracker's own `schema_version` inserts, which never get rows of their own, so "newest saved pos >= position" never becomes true (failed 5/5). The text-match barrier is correct whenever the table is non-empty, so make that precondition hold: before `SetTracking(false)`, wait for the boot-time snapshot row to exist. In the full package it's already there, so this is a single query; in an isolated run it waits ~100ms for the snapshot to land, and the reopened tracker then resumes from a saved position and replays the barrier as designed.

```go
	tsv := framework.Server
	// The reopened tracker resumes from the last row saved in _vt.schema_version.
	// If the table is still empty (the boot-time snapshot has not landed yet,
	// which is the normal case when this test runs on its own), the reopened
	// tracker instead takes a fresh snapshot at a position that may already
	// include the barrier DDL below, and the barrier is never observed. Wait
	// for at least one saved row before restarting the tracker.
	waitForSchemaVersionRow(t)
	tsv.EnableHistorian(false)
	tsv.SetTracking(false)
```

```go
// waitForSchemaVersionRow waits until the schema tracker has saved at least one
// row, so that a subsequent restart resumes from a saved position rather than
// taking a fresh snapshot.
func waitForSchemaVersionRow(t *testing.T) {
	t.Helper()
	client := framework.NewClient()
	require.Eventually(t, func() bool {
		qr, err := client.Execute("select id from _vt.schema_version limit 1", nil)
		return err == nil && len(qr.Rows) > 0
	}, 30*time.Second, 100*time.Millisecond, "schema tracker never saved its initial schema_version row")
}
```

With that applied on top of this branch: isolated `-race` runs pass 5/5 (logs show `resuming from the last saved schema version position` followed by the barrier being replayed), and the full `vttablet/endtoend` package passes under `-race` (113 tests, 0 failures, no race reports).

File: go/vt/vttablet/endtoend/vstreamer_test.go [hunk 2]
```diff
@@ -393,12 +399,29 @@ func TestSchemaVersioning(t *testing.T) {
 
 	_, err = client.Execute("drop table vitess_version", nil)
 	require.NoError(t, err)
+	_, err = client.Execute("drop table vitess_version_barrier", nil)
```
Comment (arthurschreiber) [inline go/vt/vttablet/endtoend/vstreamer_test.go]: The barrier table is only dropped if the test reaches the end. A failure earlier leaks it and the next run fails at `create table` with an unrelated-looking error. Same pattern as `vitess_version`, so feel free to ignore, but a `t.Cleanup` with `drop table if exists` right after creating it would be cheap.

File: go/vt/vttablet/endtoend/vstreamer_test.go [hunk 3]
```diff
@@ -393,12 +399,29 @@ func TestSchemaVersioning(t *testing.T) {
...
+	require.NoError(t, err)
 	_, err = client.Execute("drop table _vt.schema_version", nil)
 	require.NoError(t, err)
```
Comment (arthurschreiber) [inline go/vt/vttablet/endtoend/vstreamer_test.go]: Not from this PR, but visible in the logs from the repro: after this `drop table _vt.schema_version`, the deferred `SetTracking(true)` fails with `Table '_vt.schema_version' doesn't exist (errno 1146)` and the tracker stays dead for the rest of the process. Nothing runs after this test today, but it means `-count=2` can't work and any future test in this package that relies on the tracker would silently get nothing.

File: go/vt/vttablet/endtoend/vstreamer_test.go [hunk 4]
```diff
@@ -522,7 +545,8 @@ func validateSchemaInserted(client *framework.QueryClient, ddl string) bool {
 
 // To avoid races between ddls and the historian refreshing its cache explicitly wait for tracker's insert to be visible
 func waitForVersionInsert(client *framework.QueryClient, ddl string) (bool, error) {
-	timeout := time.After(3000 * time.Millisecond)
+	// Generous: the tracker may be saving a backlog of rows on a slow runner.
+	timeout := time.After(30 * time.Second)
```
Comment (arthurschreiber) [inline go/vt/vttablet/endtoend/vstreamer_test.go:578]: On `validateSchemaInserted` just above: pre-existing, but since this now sits under a 30s wait: `== 1` turns a duplicate row into a timeout. A duplicate is possible on the replay path when a save commits but is reported as failed (e.g. `Close()` killing the in-flight insert, which the issue mentions seeing). `>= 1` is safer. Also `qr` is nil when the query errors (table missing), so `len(qr.Rows)` panics instead of returning false.

**Review by arthurschreiber — COMMENTED** [review]
Thanks for running this down — the diagnosis of the flake matches what I see in the tracker code, and a barrier is the right shape for the fix. One blocking problem though: the barrier as written fails in isolated runs of the test, and I could reproduce that locally most of the time (details on line 80, with logs). The fix is small: the barrier is right whenever `_vt.schema_version` already has a row, so wait for the boot-time snapshot row before restarting the tracker. Patch and verification results inline.

Things I checked that look good: the "rows are saved in binlog order" argument holds (`Tracker.process` saves synchronously inside the stream callback and rewinds to `prevGtid` on error); the `vitess_version%` ACL prefix doesn't overlap any other entry, and `tableacl` would have rejected it at load if it did; and the barrier DDL round-trips through the parser byte-identical, so the text match would have worked in the state it was designed for.

One more small thing that I can't anchor inline: `getSchemaVersionTableCreationEvents()` (line 47) — with the barrier waiting on `_vt.schema_version`, the table always exists by the time this is called, so this only ever returns nil now and the comment is stale. Fine to leave for a follow-up; just noting it so nobody wonders what the 3× `gtid, other` branch is for.

File: go/vt/vttablet/endtoend/vstreamer_test.go (hunk 1 above)
Comment (mattlord) [reply go/vt/vttablet/endtoend/vstreamer_test.go]: Confirmed: every isolated run on the previous head logged `error getting the schema tracker's startup position: context canceled` followed by `Saving schema snapshot` on reopen, so the barrier was racing `engine.Reload` in `insertCurrentSchemaSnapshot`. Applied your precondition wait in 0dfa76d711 (`waitForSchemaVersionRow` before `SetTracking(false)`). Isolated `-race` runs now pass 5/5 and each reopen logs `resuming from the last saved schema version position` with the barrier streamed after it; the view-tests+no-sleep sequence that reproduced the original flake passes 4/4, and the full package passes plain and under `-race`.

[the remark this answers is not in the document]
Comment (mattlord) [reply go/vt/vttablet/endtoend/vstreamer_test.go]: Fixed in 0dfa76d711 by waiting for the boot-time snapshot row before the tracker restart, so the reopened tracker resumes from a saved position instead of snapshotting; see the thread below for the runs.

File: go/vt/vttablet/endtoend/vstreamer_test.go (hunk 2 above)
Comment (mattlord) [reply go/vt/vttablet/endtoend/vstreamer_test.go]: Done in 0dfa76d711: the barrier helper registers a `t.Cleanup` that runs `drop table if exists vitess_version_barrier`, and the explicit drop at the end of the test is gone. Verified the cleanup drop still succeeds after `_vt.schema_version` has been dropped (`schema engine ... dropped [vitess_version_barrier]` in the log).

File: go/vt/vttablet/endtoend/vstreamer_test.go (hunk 4 above)
Comment (mattlord) [reply go/vt/vttablet/endtoend/vstreamer_test.go:578]: Both in 0dfa76d711: `validateSchemaInserted` returns false on a query error instead of dereferencing a nil result, and matches `>= 1` rows with a comment on why a duplicate is possible on the replay path.

File: go/vt/vttablet/endtoend/vstreamer_test.go (hunk 3 above)
Comment (mattlord) [reply go/vt/vttablet/endtoend/vstreamer_test.go]: Agreed it's real and pre-existing: the end-of-test `drop table _vt.schema_version` leaves the deferred `SetTracking(true)` failing with errno 1146, so the tracker is dead for the rest of the process and `-count=2` can't work. I've left it out of this PR since it isn't part of the flake; happy to do it as a follow-up (likely `delete from` rather than `drop table`, now that the sidecar owns the table's creation).

Comment (mattlord) [reply go/vt/vttablet/endtoend/vstreamer_test.go]: Done in 459dddc95c, though not with `delete from` as I first suggested. Emptying the table makes the deferred restart take the snapshot path, and its `engine.Reload` raced the barrier table's cleanup DDL (`unable to get columns for table vitess_version_barrier`), which fails `startupPosition`; `process()` returns on that without retrying, so the tracker ended up just as dead, ~1 in 2 runs with `-count=2`. Leaving the rows in place makes the restart resume from the last saved position instead, which needs no reload. `-count=2 -race` now passes 3/3 with the restart logging `resuming from the last saved schema version position` and then processing the cleanup drop.

That `process()` bails permanently on a startup-position error, rather than retrying the way the stream loop does, seems worth its own issue: a transient reload failure at Open (e.g. a concurrent DDL) silently leaves a tablet without a tracker until its next state transition.

**Review by arthurschreiber — APPROVED** [review]
